### PR TITLE
Solves issue 6005 - files with same filename

### PR DESCRIPTION
**Pull request by jainsneha23**
Gets unique name with path for breakpoints

Fixes Issue: #6005

Here's the Pull Request Doc
https://firefox-dev.tools/debugger.html/docs/pull-requests.html

### Summary of Changes

* Show directories for files with same names in the breakpoint list

### Test Plan

* Open a project which has files with same filenames
* Add breakpoints in line of two files with same name, and one breakpoint in a file with unique filename
* Directory should be visible with the common filenames in the breakpoint list, and should not be visible in the file with unique filename

Example test plan:

- [x] Add breakpoints
- [x] Check breakpoint list for filenames with directories

Here's the Debugger's Testing doc
https://docs.google.com/document/d/1oBMRxV8A2ag2t22YsQOxTdEv0mXKzIg0tggJjRkU1S0/edit#.
Feel free to improve it!

### Screenshots/Videos (OPTIONAL)
screen shot 2018-07-18 at 12 05 07 pm

### CODE REVIEW

File: src/utils/source.js [hunk 1]
```diff
@@ -169,6 +169,54 @@ export function getTruncatedFileName(source: Source, length: number = 30) {
...
 }
 
+/* Show a source url's unique filename for editor tabs, breakpoints, etc.
+ * Pass the url of file, and the list of file urls
+ *
+ * @memberof utils/source
+ * @static
+ */
+
+export function getUniqueFileName(mySource: Source, sources: Source[]) {
```
Comment (jasonLaster) [inline src/utils/source.js]: lets add some unit tests for this function

File: src/utils/source.js [hunk 2]
```diff
@@ -169,6 +169,54 @@ export function getTruncatedFileName(source: Source, length: number = 30) {
...
+
+  let myPathSegments;
+  const openEditorPathSegmentsWithSameFilename = [];
+  for (const source of sources) {
+    if (!source.url) {
+      continue;
+    }
+    const fileName = getFilename(source);
+    if (fileName === myFileName) {
+      const idx = source.url.lastIndexOf("/");
```
Comment (jasonLaster) [inline src/utils/source.js]: can we use `getURL` for some of this? (path, filename)

File: src/utils/source.js [hunk 3]
```diff
@@ -169,6 +169,54 @@ export function getTruncatedFileName(source: Source, length: number = 30) {
...
+    const myPathSegment = myPathSegments[i];
+    if (
+      openEditorPathSegmentsWithSameFilename.some(
+        segments => segments.length === i + 1 || segments[i] !== myPathSegment
+      )
+    ) {
+      commonPathSegmentCount = i;
+      break;
+    }
+  }
```
Comment (jasonLaster) [inline src/utils/source.js]: this feels like something we can simplify... but i don't know how yet 😉

File: src/utils/source.js [hunk 4]
```diff
@@ -169,6 +169,54 @@ export function getTruncatedFileName(source: Source, length: number = 30) {
...
+  return `${myFileName} \u2014 ${[
+    ...myPathSegments.slice(commonPathSegmentCount + 1)
+  ].join("/")}`;
+}
```
Comment (jasonLaster) [inline src/utils/source.js]: maybe we have this just return the common paths so on the component level we can do `${myFileName} \u2014  ${commonPaths}`

**Comment by darkwing** [comment]
I *love* this!  Wow!  We should also investigate usage in tab titles.  What a great start!

File: src/utils/source.js (hunk 4 above)
Comment (jainsneha23) [reply src/utils/source.js]: nice idea. Also any nice name for the function, currently I named it 'getPath'

**Comment by jasonLaster** [comment]
One style nit, I think it'd be nice to drop the `--` and instead of the path be slightly greyer

**Comment by jainsneha23** [comment]
@jasonLaster As you recommended to drop the --, and instead the path be slightly greyer, you have any example in mind, any UX? All editors , browsers use - , using greyer text might not give a clear idea.

**Comment by jasonLaster** [comment]
Hey @jainsneha23 I think the vscode example is pretty good. 

> All editors , browsers use -

where is that?

**Comment by darkwing** [comment]
I'm so excited for this feature :D :D :D

I think one additional test that would be good would be a test that opens ~3-4 tabs, all with the same file name, but with different paths and ensure they are all as unique as they should be.

The styling of the display path can be adjusted afterward.  I think we can tie this first iteration up with @jasonLaster 's function rewrite if it works and one more test.

Amazing work @jainsneha23 !  Wow!

**Comment by jainsneha23** [comment]
@jasonLaster Please check the various styles in the image, and let me know which one to incorporate.

Also I was wondering, keeping the getDisplayPath in render function will impact performance. Shouldn't this be moved somewhere else? 

For first 2 styles in below image, there will be 2 calls for each path, to check if there is any value, if yes, render.
For other 2 styles, simple space or hypen and path, it is still fine.

screen shot 2018-07-27 at 3 36 30 pm

**Comment by jainsneha23** [comment]
Current style (Matches vscode) :

screen shot 2018-07-30 at 11 41 44 am

**Comment by jasonLaster** [comment]
Thanks so much! This is a really nice feature :)

**Pull request merged**